### PR TITLE
Lightgraph poc

### PR DESCRIPTION
**Pull request by matbesancon**
Defining the core LightGraphs functions for EvolvingGraphs types.

Big points:  
All abstract types could benefit by inheriting from the LightGraphs abtract graphs to get features for free. Lots of functions are synonyms / doubled between the two packages. Types to inherit from would be edges and graphs. 

All functions which could be defined for all EvolvingGraphs graph types are done as such. Otherwise some functions are restricted to EvolvingGraphs or StaticGraphs.

StaticGraphs can be turned into LightGraphs.SimpleGraphs, loosing the Node IDs information. We could imagine a conversion to MetaGraphs.jl to keep them.

EvolvingGraphs is a more interesting case: I think they can be represented as a path of graphs node(i,t) is linked to node(i,t +- 1), this would be better captured by MetaGraphs.jl too. Still lacking since this would have to be discussed

### CODE REVIEW

**Comment by jpfairbanks** [comment]
I think the evolving graphs team would have the expertise on how to convert a temporal graph to a static graph. I think the options are:

1. pick a time window and merge all the edges into one graph, losing temporal information. 
2. represent each (vertex, time) pair as a vertex in the new graph and then put edges between (v,t) and (v,t+1).

Are there any other options? I think that the first is often useful and the second is useful to people who know what they are doing.

**Comment by weijianzhg** [comment]
@jpfairbanks Indeed, we have implemented option 1. See `aggregate_graph`: https://etymoio.github.io/EvolvingGraphs.jl/latest/graph_types.html#EvolvingGraphs.aggregate_graph

**Comment by weijianzhg** [comment]
@mbesancon Nicely done. Thanks!

General comments: you shouldn't wrap evolving graph types for LightGraphs. Even though you can wrap things like `add_edge!`, it provides wrong results for algorithms in LightGraphs.jl. In particular, for evolving graph algorithms, we need to take care to preserve the direction of time. 

I suggest that you use `AbstractStaticGraph` rather than `AbstractGraph` and remove all wrappers for `AbstractEvolvingGraph`.

File: src/read_write/lg_interface.jl [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+Base.reverse(e::AbstractEdge{V}) where V = edge_reverse(e)
+
+LightGraphs.edges(g::AbstractGraph{V,E}) where {V,E} = edges(g)
+LightGraphs.edgetype(::AbstractGraph{V,E}) where {V,E} = E
+
+LightGraphs.vertices(g::AbstractGraph{V,E}) where {V,E} = nodes(g)
+
+LightGraphs.is_directed(g::AbstractGraph{V,E}) where {V,E} = is_directed(g)
+
+LightGraphs.add_edge!(g::AbstractStaticGraph{V}, e::AbstractEdge{ET}) where {V,ET} = add_edge!(g, LightGraphs.src(e), LightGraphs.dst(e))
```
Comment (weijianzhg) [inline src/read_write/lg_interface.jl]: I think this should be 
```
LightGraphs.add_edge!(g::AbstractStaticGraph{V,E}, e::AbstractEdge{V})
```

File: src/read_write/lg_interface.jl [hunk 2]
```diff
@@ -0,0 +1,48 @@
+LightGraphs.nv(g::AbstractGraph{V,E}) where {V,E} = num_nodes(g)
+LightGraphs.ne(g::AbstractGraph{V,E}) where {V,E} = num_edges(g)
+
+function LightGraphs.outneighbors(g::AbstractStaticGraph{NT,E}, v::V) where {V, NT<:AbstractNode{V},E}
```
Comment (weijianzhg) [inline src/read_write/lg_interface.jl]: This should be
```
LightGraphs.outneighbors(g::AbstractStaticGraph{V}, v::V)
```

File: src/read_write/lg_interface.jl [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    edges = out_edges(g, v)
+    return Node{V}[target(e) for e in edges]
+end
+
+function LightGraphs.inneighbors(g::AbstractStaticGraph{NT,E}, v::V) where {V, NT<:AbstractNode{V},E}
```
Comment (weijianzhg) [inline src/read_write/lg_interface.jl]: This should be
```
LightGraphs.inneighbors(g::AbstractStaticGraph{V}, v::V)
```

File: src/read_write/lg_interface.jl (hunk 2 above)
Comment (matbesancon) [reply src/read_write/lg_interface.jl]: I guess you will want the two:
```
LightGraphs.outneighbors(g::AbstractStaticGraph{NT,E}, v::V) where {V, NT<:AbstractNode{V},E}
```
means you can call `LightGraphs.outneighbors(g,"a")` to get neighbors for the node with key "a" while 
```
LightGraphs.outneighbors(g::AbstractStaticGraph{V,E}, v::V) where {V,E}
```
allows for `LightGraphs.outneighbors(g, nodes(g)[1])`

I added the two ways for both `in` and `out`

File: src/read_write/lg_interface.jl (hunk 1 above)
Comment (matbesancon) [reply src/read_write/lg_interface.jl]: I added the two, to be able to define an edge by keys and not indices
```
LightGraphs.add_edge!(g::AbstractStaticGraph{V,E}, e::AbstractEdge{V}) where {V,E} = add_edge!(g, LightGraphs.src(e), LightGraphs.dst(e))
LightGraphs.add_edge!(g::AbstractStaticGraph{V,E}, e::AbstractEdge{T}) where {T, V<:AbstractNode{T},E} = add_edge!(g, LightGraphs.src(e), LightGraphs.dst(e))
```

File: src/read_write/lg_interface.jl (hunk 3 above)
Comment (matbesancon) [reply src/read_write/lg_interface.jl]: fixed same way

File: src/read_write/lg_interface.jl (hunk 2 above)
Comment (weijianzhg) [reply src/read_write/lg_interface.jl]: Yes, great

File: src/read_write/lg_interface.jl [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+LightGraphs.edges(g::AbstractStaticGraph{V,E}) where {V,E} = edges(g)
+LightGraphs.edgetype(::AbstractStaticGraph{V,E}) where {V,E} = E
+
+LightGraphs.vertices(g::AbstractStaticGraph{V,E}) where {V,E} = nodes(g)
+
+LightGraphs.is_directed(g::AbstractStaticGraph{V,E}) where {V,E} = is_directed(g)
+
+LightGraphs.add_edge!(g::AbstractStaticGraph{V,E}, e::AbstractEdge{V}) where {V,E} = add_edge!(g, LightGraphs.src(e), LightGraphs.dst(e))
+LightGraphs.add_edge!(g::AbstractStaticGraph{V,E}, e::AbstractEdge{T}) where {T, V<:AbstractNode{T},E} = add_edge!(g, LightGraphs.src(e), LightGraphs.dst(e))
+LightGraphs.add_edge!(g::AbstractEvolvingGraph{V,E,T}, e::TimeEdge{ET}) where {V, E, T, ET} = add_edge!(g, LightGraphs.src(e), LightGraphs.dst(e), e.timestamp)
```
Comment (weijianzhg) [inline src/read_write/lg_interface.jl]: Please remove `LightGraphs.add_edge!(g::AbstractEvolvingGraph{V,E,T}, e::TimeEdge{ET})`

File: src/read_write/lg_interface.jl [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+LightGraphs.add_edge!(g::AbstractEvolvingGraph{V,E,T}, e::WeightedTimeEdge{EV,ET,EW}) where {V,E,T,EV,ET,EW} =
```
Comment (weijianzhg) [inline src/read_write/lg_interface.jl]: Also this one. Then I think we are ready to merge. Thanks @mbesancon !

File: src/read_write/lg_interface.jl (hunk 4 above)
Comment (matbesancon) [reply src/read_write/lg_interface.jl]: removed in last commit

**Comment by sbromberger** [comment]
Just out of curiosity, does something like `betweenness_centrality` now work for EvolvingGraphs with this setup? (I use this metric as an advanced test of API functionality.)

**Comment by weijianzhg** [comment]
Hi @sbromberger, I am afraid it won't work. Betweenness centrality is based on the definition of shortest path. In evolving graph, we need to use shortest temporal path which is different from the static case.

**Pull request merged**